### DISCUSSION TITLE
Insert ignore statements with ALE?

### DISCUSSION THREAD

**Discussion by Erotemic** [w0rp/ale#4489] [Q&A - Ask for help with problems]
I'm wondering if there exists (or is any any easy way to implement) a feature where I can insert an "ignore directive" for particular lint warnings.

The specific use case I have is shellcheck. It seems to update itself with new sets of warnings all the time, and while this is useful it often gives me warning for things I'm aware are ok in the circumstance.

I currently just go through them line by line and fix them or add a `    # shellcheck disable=<num>` line above the offending line. Similarly with Python I would write a  `# NOQA` (which is a bit easier because I don't need to write out the error code).

Is there a way that ale can insert the appropriate ignore directive into the code? If not, is there a way that I could implement a function in my vimrc that inspects the current linting error and grabs its error code so I can insert the appropriate text?

**Comment by hsanson** [comment]
In short: no.

The principle is that the editor itself should not hide warnings or errors reported by the underlining tools (linters). If this is necessary for the specific project it should be done by the underlining tool itself so everyone working on the same project get the same errors and warnings idependently of the editor they use.

Related discussions:
 - https://github.com/dense-analysis/ale/issues/4385#issuecomment-1415107370
 - https://github.com/dense-analysis/ale/issues/3465#issuecomment-1416178649

**Reply by Erotemic** [reply]
I think there is a misunderstanding. I'm not trying to get ALE to ignore the warning. I'm trying to get ALE to generate the text that tells the tool to ignore the warnings.

E.g. If I do something like:


```python
import unused
```

I want to some shortcut to generate and insert the text:

```python
import unused  # NOQA: F401
```

**Reply by hsanson** [reply]
I see...  I think same principle applies.  ALE just exposes whatever the underlying tools offers and combines them in a consistent interface to VIM/Neovim. 

Inspecting the list of reported errors and manipulating the buffer based on them should be possible.  You can see here for an example on how to retrieve list of all issues reported by ALE for all open buffers:

- https://github.com/hsanson/nerdtree-ale/blob/master/nerdtree_plugin/ale.vim#L36

First it uses `ALELintPost` event to trigger a function. This ensures this is called after ALE has finished linting and filling the list of errors.

Then uses `ale#engine#GetLoclist(l:buffer)`  to get the list of reported issues. It should be possible to call this on the current buffer only and iterate over each error to check them and update the buffer contents accordingly.

Certainly not an easy task but not impossible I would say.

**Reply by Erotemic** [reply]
Thanks, I think GetLocList can get me started. 

Say I have some setting like this:

![image](https://user-images.githubusercontent.com/3186211/228872524-710460d7-d485-4d59-ba28-1d4c8da2ec9b.png)


If I have my cursor over 1) the line in the main buffer containiner the error or 2) the line in the location list reporting the error, for me to ask ALE for details about the error? In the second case I could probably just parse the line (annoying but doable), but in the first case, can I ask ALE what error corresponds to a line? Or given a line do I have to search LocList for the set of errors that have the same line?

**Reply by hsanson** [reply]
ALE has no such method but you should be able to use getqflist() or getloclist() depending on which one you set ALE to use to get specific errors.  Read the vim help for more details.
